src/Microsoft.ML.Dnn/ImageClassificationTransform.cs

```diff
@@ -749,6 +807,12 @@ private protected override void SaveModel(ModelSaveContext ctx)
             ctx.Writer.Write(_predictedLabelColumnName);
             ctx.Writer.Write(_learningRate);
             ctx.Writer.Write(_classCount);
+
+            Host.AssertNonEmpty(_keyValueAnnotations);
+            Host.Assert(_keyValueAnnotations.Length == _classCount);
+            for (int j = 0; j < _classCount; j++)
```

Comment:
Just a minor FYI - 

In C#, when you index into an array, the index is bounds-checked at runtime. The runtime checks if the index is non-negative, and less than the length of the array. If it is outside those bounds, an exception is thrown. While these bounds checks aren't very expensive, they can add up.
The C# compiler is smart enough to notice this pattern:

```C#
int[] myArray = ...;
for (int i = 0; i < myArray.Length; i++)
{
    int element = myArray[i];
}
```

It sees this pattern, and knows that `i` will never be outside the bounds of `myArray`. So the compiler removes the bounds checks as a performance improvement.

The take away here is, when you have 2 variables to pick from to use in your `for` loop break condition, and one of them is the `.Length` of an array you are indexing into, and one isn't. Choose the `array.Length` variable, so the bounds checks will be removed.

Reply:
I see, that makes sense, thank you Eric! It's pretty cool that the C# compiler is able to recognize the traversing of an array and modify the bound checks as it goes.